src/token/cose/encrypt/encrypt0.rs

```diff
@@ -0,0 +1,553 @@
+    unprotected: Option<&Header>,
+    try_all_keys: bool,
+    plaintext: &[u8],
+    // NOTE: aad ist not the external AAD provided by the user, but the Enc_structure as defined in RFC 9052, Section 5.3
+    aad: &[u8],
+) -> Result<Vec<u8>, CoseCipherError<B::Error>> {
+    let parsed_key = determine_key_candidates(
+        key_provider,
+        protected,
+        unprotected,
+        &KeyOperation::Assigned(iana::KeyOperation::Sign),
+        false,
+    )?
+    .into_iter()
+    .next()
+    .ok_or(CoseCipherError::NoKeyFound)?;
+    let algorithm = determine_algorithm(&parsed_key, protected, unprotected)?;
+
+    match algorithm {
+        Algorithm::Assigned(
+            iana::Algorithm::A128GCM | iana::Algorithm::A192GCM | iana::Algorithm::A256GCM,
+        ) => {
+            // Check if this is a valid AES key.
+            let symm_key = is_valid_aes_key::<B::Error>(&algorithm, parsed_key)?;
+
+            let iv = if protected.is_some() && !protected.unwrap().iv.is_empty() {
+                protected.unwrap().iv.as_ref()
+            } else if unprotected.is_some() && !unprotected.unwrap().iv.is_empty() {
+                unprotected.unwrap().iv.as_ref()
+            } else {
+                return Err(CoseCipherError::IvRequired);
+            };
+
+            backend.encrypt_aes_gcm(algorithm, symm_key, plaintext, aad, iv)
+        }
+        v @ (Algorithm::Assigned(_)) => Err(CoseCipherError::UnsupportedAlgorithm(v.clone())),
+        // TODO make this extensible? I'm unsure whether it would be worth the effort, considering
+        //      that using your own (or another non-recommended) algorithm is not a good idea anyways.
+        v @ (Algorithm::PrivateUse(_) | Algorithm::Text(_)) => {
+            Err(CoseCipherError::UnsupportedAlgorithm(v.clone()))
+        }
+    }
+}
+
+fn try_decrypt_with_key<B: CoseEncryptCipher>(
+    backend: &mut B,
+    key: CoseParsedKey<B::Error>,
+    protected: &Header,
+    unprotected: &Header,
+    ciphertext: &[u8],
+    // NOTE: aad ist not the external AAD provided by the user, but the Enc_structure as defined in RFC 9052, Section 5.3
+    aad: &[u8],
+) -> Result<Vec<u8>, CoseCipherError<B::Error>> {
+    let algorithm = determine_algorithm(&key, Some(protected), Some(unprotected))?;
+
+    match algorithm {
+        Algorithm::Assigned(
+            iana::Algorithm::A128GCM | iana::Algorithm::A192GCM | iana::Algorithm::A256GCM,
+        ) => {
+            // Check if this is a valid AES key.
+            let symm_key = is_valid_aes_key::<B::Error>(&algorithm, key)?;
+
+            let iv = if !protected.iv.is_empty() {
+                protected.iv.as_ref()
+            } else if !unprotected.iv.is_empty() {
+                unprotected.iv.as_ref()
+            } else {
+                return Err(CoseCipherError::IvRequired);
+            };
+
+            // Authentication tag is 16 bytes long and should be included in the ciphertext.
+            if ciphertext.len() < 16 {
```

Comment:
Again, just to check: Do we allow empty ciphertexts (after subtracting the authentication tag)?

Reply:
> There should probably be a test for that.

Agreed. If this is for some reason more effort than it seems, we could also do this in a follow-up issue.